Specify the rendering of <details> in more detail.

dbaron opened the pull request:
This makes a number of related changes to specify the rendering of the `<details>` element:

* It specifies the structure of the user agent shadow tree.  This appears largely interoperable between implementations with the exception of the style or link element for the default summary styles: Gecko uses a link element as the first child, Chromium uses a style element as the last child, and WebKit does not use a style element (see below).  This specifies a style or link as the last child with a note that the order is not observable.
* It specifies the existence of the default summary and the presence of UA dependent text inside of it.  This is present in all of Chromium, Gecko, and WebKit.
* It specifies the styles needed for the default summary.  These match Chromium and Gecko.  (These are not present in WebKit because WebKit's mechanism for styling the marker does not match the existing spec; see [bug 157323](https://bugs.webkit.org/show_bug.cgi?id=157323).)
* It removes the restriction that `<details>` is a block and cannot be changed.  This is prototyped in Chromium and Gecko.  This fixes #9830.
* It specifies that the summary element is display: block by default. This matches all of Chromium, Gecko, and WebKit.
* It specifies which element matches the `::details-content` pseudo-element.  This is prototyped in Chromium.
* It makes a potentially riskier change by making the `slot` element that matches `::details-content` be `display: block` all the time (rather than only when closed).  This is prototyped in Chromium (more recently than the other changes).

See https://github.com/w3c/csswg-drafts/issues/9879 and https://github.com/w3c/csswg-drafts/issues/9951 for more background.

<!--
Thank you for contributing to the HTML Standard! Please describe the change you are making and complete the checklist below if your change is not editorial.
When editing this comment after the PR is created, check that PR-Preview doesn't overwrite your changes.
If you think your PR is ready to land, please double-check that the build is passing and the checklist is complete before pinging.
-->

- [X] At least two implementers are interested (and none opposed):
   * Chromium *support*, see [issue 40068507](https://issues.chromium.org/issues/40068507)
   * Gecko *position pending*: shown partial interest based on [bug 1856374](https://bugzilla.mozilla.org/show_bug.cgi?id=1856374) but that's only a piece of this; remainder is covered by https://github.com/mozilla/standards-positions/issues/1027
   * WebKit *support*: https://github.com/WebKit/standards-positions/issues/351
- [ ] [Tests](https://github.com/web-platform-tests/wpt) are written and can be reviewed and commented upon at:
   * https://wpt.fyi/results/html/rendering/the-details-element/details-display-type-001.tentative.html
   * https://wpt.fyi/results/html/rendering/the-details-element/details-display-type-002.tentative.html
   * https://wpt.fyi/results/html/rendering/the-details-element/details-pseudo-elements-001.tentative.html
   * https://wpt.fyi/results/html/rendering/the-details-element/details-pseudo-elements-002.tentative.html
   * https://wpt.fyi/results/html/rendering/the-details-element/details-pseudo-elements-003.tentative.html
   * https://wpt.fyi/results/html/rendering/the-details-element/details-display.tentative.html (just added by https://github.com/web-platform-tests/wpt/pull/46844)
   * PR to make non-tentative: https://github.com/web-platform-tests/wpt/pull/46847
- [X] [Implementation bugs](https://github.com/whatwg/meta/blob/main/MAINTAINERS.md#handling-pull-requests) are filed:
   * Chromium: [issue 40068507](https://issues.chromium.org/issues/40068507)
   * Gecko: [bug 1901037](https://bugzilla.mozilla.org/show_bug.cgi?id=1901037)
   * WebKit: [bug 275211](https://bugs.webkit.org/show_bug.cgi?id=275211)
- [X] [MDN issue](https://github.com/whatwg/meta/blob/main/MAINTAINERS.md#handling-pull-requests) is filed: https://github.com/mdn/mdn/issues/559
- [X] The top of this comment includes a [clear commit message](https://github.com/whatwg/meta/blob/main/COMMITTING.md) to use. <!-- If you created this PR from a single commit, Github copied its message. Otherwise, you need to add a commit message yourself. -->

(See [WHATWG Working Mode: Changes](https://whatwg.org/working-mode#changes) for more details.)


<!--
    This comment and the below content is programmatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://whatpr.org/html/10265/infrastructure.html" title="Last updated on Jun 20, 2024, 5:39 PM UTC (4abf496)">/infrastructure.html</a>  ( <a href="https://whatpr.org/html/10265/11ea4a2...4abf496/infrastructure.html" title="Last updated on Jun 20, 2024, 5:39 PM UTC (4abf496)">diff</a> )
<a href="https://whatpr.org/html/10265/rendering.html" title="Last updated on Jun 20, 2024, 5:39 PM UTC (4abf496)">/rendering.html</a>  ( <a href="https://whatpr.org/html/10265/11ea4a2...4abf496/rendering.html" title="Last updated on Jun 20, 2024, 5:39 PM UTC (4abf496)">diff</a> )

dbaron commented:
Sorry, apparently Shift-enter submits the form, I wasn't quite ready to submit and I'll continue editing the above comment.

domenic commented:
> * It specifies the structure of the user agent shadow tree.  This appears largely interoperable between implementations with the exception of the style or link element for the default summary styles: Gecko uses a link element as the first child, Chromium uses a style element as the last child, and WebKit does not use a style element (see below).  This specifies a style or link as the first child.

Are these differences observable? If not, then perhaps the spec can be clearer about that, so that people understand any implementation strategy that achieves equivalent results is fine.

dbaron commented:
> Are these differences observable? If not, then perhaps the spec can be clearer about that, so that people understand any implementation strategy that achieves equivalent results is fine.

That's a good question.  When I wrote it I was (for some reason) thinking that the ordering was observable via selectors, but now that I think about it again I think it's not.  We do need to get the exact CSS definitions of what does and doesn't apply sorted out, though, to make sure.  But I think as long as (a) `link` or `style` in the shadow tree are `display:none` and can't be changed and (b) the relative positions of the elements in the shadow tree can't be determined by selectors, then I think it's not observable.  (The relative positions of displayable elements in the shadow tree can be determined by layout differences, probably most easily with `display: grid`.)

annevk commented on source:
```diff
@@ -134178,28 +134181,53 @@ details[open] > summary:first-of-type {
+    <p>The second child element is a <code>slot</code> that is expected to take the
+    <code>details</code> element's first <code>summary</code> element child, if any.  This element
+    has a single child <code>summary</code> element called the <dfn>default summary</dfn> which has
+    text content that is <span>implementation-defined</span> (and probably locale-specific).</p>
```
Should we recommend here that it follows from the node's language? I think that's what we do for form controls today even though it's not (yet) implemented?

dbaron replied:
This matches the current wording for the default text of submit/reset buttons and (except for "possibly" versus "probably") the text of the "Choose file" button in `<input type=file>`.  The only wording I found that seems like what you're thinking of was the [wording about presentation of dates, times, and numbers](https://html.spec.whatwg.org/multipage/input.html#input-impl-notes).  This seems more similar to submit/reset and to the file chooser button than it it does to presentation of dates, in that it feels a bit odd to make this *differ* from those.

annevk replied:
I suspect that when we changed the wording for dates, times, and numbers, the button language was overlooked. But I guess we can sort that out in a follow-up.

dbaron replied:
Filed #10437 to follow up.

annevk commented on source:
```diff
@@ -134178,28 +134181,53 @@ details[open] > summary:first-of-type {
+    <p>The third child element is a <code>slot</code> that is expected to take the
+    <code>details</code> element's remaining descendants, if any.  This element has no contents.</p>
+
+    <!-- TODO: link to CSS definition when spec is ready -->
+    <p>This element is expected to match the '::details-content' pseudo-element.</p>
```
Should this also be mentioned in HTML's Selectors section?

dbaron replied:
I'm not sure.  The one existing pseudo-element (`::file-selector-button`) is not currently mentioned [there](https://html.spec.whatwg.org/multipage/semantics-other.html#selectors).  It seems like it might be useful to mention the pseudo-elements in that section, but it also seems like maybe (unlike for the pseudo-classes) if we do so, perhaps the pseudo-elements part of that section should be more like an index than the location of the formal definition?

annevk replied:
Yeah, I guess that's correct. In which case maybe it's not worth it.

annevk reviewed (commented):
Can we describe it as two child elements and an associated style sheet? It seems like it should be immaterial how the style sheet is associated with the shadow tree. It just needs to have one.

dbaron commented:
> Are these differences observable? If not, then perhaps the spec can be clearer about that, so that people understand any implementation strategy that achieves equivalent results is fine.

> Can we describe it as two child elements and an associated style sheet? It seems like it should be immaterial how the style sheet is associated with the shadow tree. It just needs to have one.

I've revised the change to go mostly in this direction:  I still describe the style as an element, explicitly note that the order isn't observable, and note that its association using an element isn't observable either (although I don't know what other standard mechanism we'd describe the styles with).  I also now describe it as the third child rather than the first, since in hindsight it seems nice to use "first" and "second" for the first and second rendered children.  But I didn't go so far as to just handwave about how the association is made.  Does that seem reasonable?

domenic reviewed (approved):
Editorially LGTM. Happy to merge when we have the checkboxes filled out. Let us know if you have other people (e.g. @emilio) whose reviews you want to block on before merging.

emilio commented:
> I've revised the change to go mostly in this direction: I still describe the style as an element, explicitly note that the order isn't observable, and note that its association using an element isn't observable either (although I don't know what other standard mechanism we'd describe the styles with).

Adopted stylesheets would be an option fwiw.

> I also now describe it as the third child rather than the first, since in hindsight it seems nice to use "first" and "second" for the first and second rendered children. But I didn't go so far as to just handwave about how the association is made. Does that seem reasonable?

Works for me.

emilio commented on source:
```diff
@@ -134178,28 +134181,62 @@ details[open] > summary:first-of-type {
+   <li>
+    <p>The third child element is either a <code>link</code> or <code>style</code>
+    element with the following styles for the <span>default summary</span>:</p>
+    <pre><code class="css">:host summary {
```
You can probably use `:host > summary` for parallelism with the rule below right? (or just `summary`?)

domenic replied:
@dbaron, ping to make sure you saw this, as I agree it seems strange for these two lines to diverge.

dbaron replied:
Sorry for missing this.  I've now made them consistent by changing the second selector, since that matches both [Chrome's code](https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/core/html/html_details_element.cc;l=117-126;drc=bfe80d0c3bd0e03e03a829138685704a28b13708) and [Firefox's code](https://searchfox.org/mozilla-central/rev/74518d4f6979b088e28405ba7e6238f4707639cf/layout/style/res/details.css#14-21).

domenic replied:
Awesome. Can you be sure this distinction is tested?

dbaron replied:
I think it's untestable, since it's styling the UA shadow DOM, and the only `summary` in that shadow DOM is in fact a child of the host.

(But when I was adjusting for consistency it seemed to make sense to match the existing implementations even when the difference didn't matter.)

emilio replied:
Yeah it's not, because it's the slot fallback.

dbaron commented:
One substantive thing that I'm thinking of changing here:  right now the `::details-content` is a pseudo-element that addresses an element that is internally a `<slot>`.  `<slot>` elements default to `display: contents`.  This is somewhat inconvenient since some styles don't work on elements that are `display: contents`.  So I think probably it should be `display: block` by default all of the time (rather than only when it's hidden).

(This has tripped me up multiple times while writing examples, and I think it's probably better to change it if it doesn't break anything, which I don't *think* it will.)

bramus commented:
> I think probably it should be `display: block` by default all of the time (rather than only when it's hidden).

In all my experiments ([here](https://codepen.io/bramus/pen/ZEPzzJq), [here](https://codepen.io/bramus/pen/LYaPPGK), and [here](https://codepen.io/bramus/pen/bGZbGZP)) I have done this, so +1 on using `display: block` by default.

dbaron commented:
I think this is ready for another look.  There are a few additional changes since the last round of reviews; the draft commit message above is updated to reflect them.

(There's one checkbox still unchecked, but that's just because I'm waiting to make a PR to make the tests not tentative until the last tentative test lands, hopefully later today or Thursday.)

domenic reviewed (approved):
LGTM, modulo last question about more tests.

It looks like since this last got a detailed look from @annevk and @emilio, we changed ::details-content to be `display: block`. (Is that tested too?)

So, I'll give them a few more days to chime in, and plan on merging this around Wednesday Japan time if nothing further comes in.

emilio commented on source:
```diff
@@ -134178,28 +134182,62 @@ details[open] > summary:first-of-type {
+    <p>This element is expected to match the <span>'::details-content'</span> pseudo-element.</p>
+
+    <p>This element is expected to have its <code data-x="attr-style">style</code> attribute set to
+    "<code data-x="">display: block; content-visibility: hidden;</code>" when the
```
I think it'd be slightly nicer to make the display: block change part of the UA sheet. Maybe `slot + slot { display: block }`? Same with the content-visibility one for what is worth (we do [this](https://searchfox.org/mozilla-central/rev/cb1060f7b4581e6c2d30f1accc84c7d807132d82/layout/style/res/details.css#5-11)), but I guess it doesn't matter much and it's not really observable. It's just less magic

dbaron replied:
My inclination is that it is probably reasonable to do that as a separate change, but I'd prefer not to make this change bigger (this is an issue that already exists prior to this PR), and I'd probably also like to check it by implementing it and making sure it actually works the way we think it does.

emilio replied:
Sure, sorry should've clarified that my comment wasn't meant to be blocking in any way. Sounds good. Maybe file a follow-up issue for this?

dbaron replied:
Filed #10438 to follow up.
